src/refract/JsonSchema.cc

```diff
@@ -0,0 +1,290 @@
+
+    template <typename... Args>
+    so::Object& add_enum(so::Object& schema, Args&&... args)
+    {
+        schema.data.emplace_back("enum", so::Array{ so::Value{ std::forward<Args>(args) }... });
+        return schema;
+    }
+}
+
+namespace
+{
+    so::Object& render_schema(so::Object& schema, const ObjectElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const ArrayElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const EnumElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const NullElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const StringElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const NumberElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const BooleanElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const MemberElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const ExtendElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const OptionElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const SelectElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const RefElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const HolderElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, const IElement& e, flags options = 0);
+
+    so::Object make_schema(const IElement& e, flags options = 0);
+
+    so::Object& render_schema(so::Object& schema, const ObjectElement& e, flags options)
+    {
+        assert(false); // TODO @tjanc@ not implemented
+        return schema;
+    }
+
+    so::Object& render_schema(so::Object& schema, const ArrayElement& e, flags options)
+    {
+        add_type(schema, "array");
+
+        // UPDATE OPTIONS
+        if (has_fixed_attr(e))
+            options.set(fixed_flag);
+
+        if (has_fixed_type_attr(e)) { // array of any of types
+
+            if (e.empty() || e.get().empty())
+                add_enum(schema, so::Array{});
+
+            else if (e.get().size() == 1)
+                add_items(schema, make_schema(*e.get().begin()[0], options));
+
+            else {
+                so::Array items{};
+                for (const auto& entry : e.get())
+                    items.data.emplace_back(make_schema(*entry, options));
+                add_items(schema, so::Object{ std::make_pair("anyOf", std::move(items)) });
+            }
+
+        } else if (options.test(fixed_flag)) { // tuple of constants/types
+
+            if (e.empty() || e.get().empty())
+                add_enum(schema, so::Array{});
+            else {
+                so::Array items{};
+                for (const auto& item : e.get())
+                    items.data.emplace_back(make_schema(*item, options));
+                add_items(schema, std::move(items));
```

Comment:
Missing `"minItems":<N>' and `"additionalItems":false` here. Will be fixed in following PR.